src/core/blocks/serialization.cpp

```diff
@@ -385,9 +387,75 @@ struct WritePNG : public FileBase {
     int w = int(input.payload.imageValue.width);
     int h = int(input.payload.imageValue.height);
     int c = int(input.payload.imageValue.channels);
-    if (0 == stbi_write_png(filename.c_str(), w, h, c,
-                            input.payload.imageValue.data, w * c))
-      throw ActivationError("Failed to write PNG file.");
+
+    // demultiply alpha if needed, limited to 4 channels
+    if (c == 4 &&
+        (input.payload.imageValue.flags & CBIMAGE_FLAGS_PREMULTIPLIED_ALPHA) ==
+            CBIMAGE_FLAGS_PREMULTIPLIED_ALPHA) {
+      _scratch.resize(w * h * 4 * pixsize);
+      int lineSize = w * 4;
+      for (int ih = 0; ih < h; ih++) {
+        for (int iw = 0; iw < w; iw++) {
```

Comment:
Is it done in two nested loops for better readability or is there another reason?

Reply:
Just simplicity, I think any half decent compiler will optimize properly, but not to the extent of packing into 32 bit int.
The thing is that this block is slow in any case, we don't really need to optimize them yet.
Using (Await (WritePNG)) is advised (will run on a worker thread)

Reply:
If I pack in 32 bits then 16bit and 32bit should be done with vectors :D - let's just add comments about it for really when there is time to waste